PhysicsTools/XGBoost/src/XGBooster.cc

```diff
@@ -69,28 +71,41 @@ XGBooster::XGBooster(std::string model_file, std::string model_features) : XGBoo
   }
 }
 
-void XGBooster::reset() { std::fill(features_.begin(), features_.end(), std::nan("")); }
+void XGBooster::reset() { std::fill(status_.begin(), status_.end(), false); }
 
 void XGBooster::addFeature(std::string name) {
   features_.push_back(0);
+  status_.push_back(false);
   feature_name_to_index_[name] = features_.size() - 1;
 }
 
-void XGBooster::set(std::string name, float value) { features_.at(feature_name_to_index_[name]) = value; }
+void XGBooster::set(std::string name, float value) {
+  auto i = feature_name_to_index_[name];
+  features_.at(i) = value;
+  status_.at(i) = true;
+}
+
+const std::string& XGBooster::name(unsigned int feature) const {
+  for (const auto& pair : feature_name_to_index_) {
+    if (pair.second == feature)
+      return pair.first;
+  }
+  throw std::runtime_error("Invalid feature index");
+}
 
 float XGBooster::predict(const int iterationEnd) {
-  // check if all feature values are set properly
-  for (unsigned int i = 0; i < features_.size(); ++i)
+  // Validate input
+  for (unsigned int i = 0; i < features_.size(); ++i) {
+    // Check that feature is set
+    if (not status_.at(i)) {
+      throw std::runtime_error("Feature " + name(i) + " not set");
+    }
+    // Check for invalid input
     if (std::isnan(features_.at(i))) {
-      std::string feature_name;
-      for (const auto& pair : feature_name_to_index_) {
-        if (pair.second == i) {
-          feature_name = pair.first;
-          break;
-        }
-      }
-      throw std::runtime_error("Feature is not set: " + feature_name);
+      edm::LogWarning("InvalidInput") << "Invalid input is provided (NaN). Feature name " + name(i);
+      return -998.;
```

Comment:
I think it's fine to raise an exception is one input is not set, and avoid the execution if a feature is NaN. 

However the special return value -998. is rather "special" and it may be missed by caller code.  Before this PR there was an exception, now it may go unchecked (and have physics effects?). I think we should include in this PR also checks on the predict output in the producers: each caller may decide to raise an exception depending on the case.

Reply:
We either don't tolerated NaNs everywhere or we tolerate them everywhere. I had to introduce the exception in the original version because XGBoost was not tolerating NaNs, while in CMSSW it seems to be a normal practice.